--- ../alsa-kernel/usb/helper.c	2011-09-27 09:22:10.436753250 +0200
+++ helper.c	2011-09-27 09:24:21.577813168 +0200
@@ -1,3 +1,5 @@
+#define __NO_VERSION__
+#include "usbaudio.inc"
 /*
  *   This program is free software; you can redistribute it and/or modify
  *   it under the terms of the GNU General Public License as published by
@@ -91,8 +93,13 @@
 		if (!buf)
 			return -ENOMEM;
 	}
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 12)
 	err = usb_control_msg(dev, pipe, request, requesttype,
 			      value, index, buf, size, 1000);
+#else
+	err = usb_control_msg(dev, pipe, request, requesttype,
+			      value, index, buf, size, HZ);
+#endif
 	if (size > 0) {
 		memcpy(data, buf, size);
 		kfree(buf);
