diff -u /tmp/rdesktop-1.8.0/channels.c /tmp/rdesktop-1.8.0-usbrdr/channels.c
--- /tmp/rdesktop-1.8.0/channels.c	2013-06-20 15:56:15.000000000 +0300
+++ /tmp/rdesktop-1.8.0-usbrdr/channels.c	2014-06-16 19:55:04.630098925 +0300
@@ -85,6 +85,7 @@
 #ifdef WITH_SCARD
 	scard_lock(SCARD_LOCK_CHANNEL);
 #endif
+	usbrdr_lock(USBRDR_LOCK_CHANNEL);
 
 	/* first fragment sent in-place */
 	s_pop_layer(s, channel_hdr);
@@ -129,6 +130,7 @@
 		data += thislength;
 	}
 
+	usbrdr_unlock(USBRDR_LOCK_CHANNEL);
 #ifdef WITH_SCARD
 	scard_unlock(SCARD_LOCK_CHANNEL);
 #endif
diff -u /tmp/rdesktop-1.8.0/constants.h /tmp/rdesktop-1.8.0-usbrdr/constants.h
--- /tmp/rdesktop-1.8.0/constants.h	2013-06-20 15:56:15.000000000 +0300
+++ /tmp/rdesktop-1.8.0-usbrdr/constants.h	2014-06-16 19:55:04.630098925 +0300
@@ -487,6 +487,12 @@
 #define SCARD_LOCK_RDPDR	3
 #define SCARD_LOCK_LAST		4
 
+/* USBRDR constants */
+#define USBRDR_LOCK_TCP		0
+#define USBRDR_LOCK_SEC		1
+#define USBRDR_LOCK_CHANNEL	2
+#define USBRDR_LOCK_RDPDR	3
+#define USBRDR_LOCK_LAST	4
 
 /* redirect flags, from [MS-RDPBCGR] 2.2.13.1 */
 enum RDP_PDU_REDIRECT_FLAGS
diff -u /tmp/rdesktop-1.8.0/Makefile.in /tmp/rdesktop-1.8.0-usbrdr/Makefile.in
--- /tmp/rdesktop-1.8.0/Makefile.in	2013-04-17 16:44:15.000000000 +0300
+++ /tmp/rdesktop-1.8.0-usbrdr/Makefile.in	2014-06-16 19:55:04.626098925 +0300
@@ -16,7 +16,7 @@
 
 CC          = @CC@
 INSTALL     = @INSTALL@
-CFLAGS      = @CFLAGS@ @X_CFLAGS@ @DEFS@ -DKEYMAP_PATH=\"$(KEYMAP_PATH)\"
+CFLAGS      = @CFLAGS@ @X_CFLAGS@ @DEFS@ -DKEYMAP_PATH=\"$(KEYMAP_PATH)\" -pthread
 LDFLAGS     = @LDFLAGS@ @LIBS@ @X_LIBS@ @X_EXTRA_LIBS@
 STRIP       = @STRIP@
 
@@ -28,7 +28,7 @@
 SCARDOBJ    = @SCARDOBJ@
 CREDSSPOBJ  = @CREDSSPOBJ@
 
-RDPOBJ   = tcp.o asn.o iso.o mcs.o secure.o licence.o rdp.o orders.o bitmap.o cache.o rdp5.o channels.o rdpdr.o serial.o printer.o disk.o parallel.o printercache.o mppc.o pstcache.o lspci.o seamless.o ssl.o utils.o
+RDPOBJ   = tcp.o asn.o iso.o mcs.o secure.o licence.o rdp.o orders.o bitmap.o cache.o rdp5.o channels.o rdpdr.o serial.o printer.o disk.o parallel.o printercache.o mppc.o pstcache.o lspci.o usbrdr.o seamless.o ssl.o utils.o
 X11OBJ   = rdesktop.o xwin.o xkeymap.o ewmhints.o xclip.o cliprdr.o ctrl.o
 VNCOBJ   = vnc/rdp2vnc.o vnc/vnc.o vnc/xkeymap.o vnc/x11stubs.o
 
@@ -85,7 +85,7 @@
 	bitmap.c cache.c channels.c cliprdr.c disk.c mppc.c ewmhints.c	\
 	iso.c licence.c mcs.c orders.c parallel.c printer.c printercache.c \
 	pstcache.c rdesktop.c rdp5.c rdp.c rdpdr.c rdpsnd.c \
-	secure.c serial.c tcp.c xclip.c xkeymap.c xwin.c lspci.c seamless.c \
+	secure.c serial.c tcp.c xclip.c xkeymap.c xwin.c lspci.c usbrdr.c seamless.c \
 	scard.c >> proto.h
 	cat proto.tail >> proto.h
 
diff -u /tmp/rdesktop-1.8.0/parse.h /tmp/rdesktop-1.8.0-usbrdr/parse.h
--- /tmp/rdesktop-1.8.0/parse.h	2013-06-20 15:56:15.000000000 +0300
+++ /tmp/rdesktop-1.8.0-usbrdr/parse.h	2014-06-16 19:55:04.630098925 +0300
@@ -32,7 +32,7 @@
 	unsigned char *sec_hdr;
 	unsigned char *rdp_hdr;
 	unsigned char *channel_hdr;
-
+	unsigned char occupied;	
 }
  *STREAM;
 
diff -u /tmp/rdesktop-1.8.0/proto.h /tmp/rdesktop-1.8.0-usbrdr/proto.h
--- /tmp/rdesktop-1.8.0/proto.h	2013-06-20 15:56:15.000000000 +0300
+++ /tmp/rdesktop-1.8.0-usbrdr/proto.h	2014-06-16 19:55:04.630098925 +0300
@@ -333,6 +333,8 @@
 void ui_seamless_ack(unsigned int serial);
 /* lspci.c */
 RD_BOOL lspci_init(void);
+/* usbrdr.c */
+RD_BOOL usbrdr_init(void);
 /* seamless.c */
 RD_BOOL seamless_init(void);
 unsigned int seamless_send_sync(void);
@@ -352,6 +354,9 @@
 void scardSetInfo(uint32 device, uint32 id, uint32 bytes_out);
 void scard_reset_state();
 
+void usbrdr_lock(int lock);
+void usbrdr_unlock(int lock);
+
 /* *INDENT-OFF* */
 #ifdef __cplusplus
 }
diff -u /tmp/rdesktop-1.8.0/rdesktop.c /tmp/rdesktop-1.8.0-usbrdr/rdesktop.c
--- /tmp/rdesktop-1.8.0/rdesktop.c	2013-06-27 13:37:42.000000000 +0300
+++ /tmp/rdesktop-1.8.0-usbrdr/rdesktop.c	2014-06-16 19:55:04.634098925 +0300
@@ -102,6 +102,7 @@
 RD_BOOL g_console_session = False;
 RD_BOOL g_numlock_sync = False;
 RD_BOOL g_lspci_enabled = False;
+RD_BOOL g_usbrdr_enabled = False;
 RD_BOOL g_owncolmap = False;
 RD_BOOL g_ownbackstore = True;	/* We can't rely on external BackingStore */
 RD_BOOL g_seamless_rdp = False;
@@ -244,6 +245,8 @@
 	fprintf(stderr,
 		"                   \"AKS\"              -> Device vendor name                 \n");
 #endif
+	fprintf(stderr,
+		"         '-r usbrdr': enable USB redirection\n");
 	fprintf(stderr, "   -0: attach to console\n");
 	fprintf(stderr, "   -4: use RDP version 4\n");
 	fprintf(stderr, "   -5: use RDP version 5 (default)\n");
@@ -834,6 +837,10 @@
 				{
 					g_lspci_enabled = True;
 				}
+				else if (str_startswith(optarg, "usbrdr"))
+				{
+					g_usbrdr_enabled = True;
+				}
 				else if (str_startswith(optarg, "lptport"))
 				{
 					parallel_enum_devices(&g_num_devices, optarg + 7);
@@ -873,7 +880,7 @@
 				}
 				else
 				{
-					warning("Unknown -r argument\n\n\tPossible arguments are: comport, disk, lptport, printer, sound, clipboard, scard\n");
+					warning("Unknown -r argument\n\n\tPossible arguments are: comport, disk, lptport, printer, sound, clipboard, scard, usbrdr\n");
 				}
 				break;
 
@@ -1084,6 +1091,9 @@
 	if (g_lspci_enabled)
 		lspci_init();
 
+	if (g_usbrdr_enabled)
+		usbrdr_init();
+
 	rdpdr_init();
 	g_reconnect_loop = False;
 	while (1)
diff -u /tmp/rdesktop-1.8.0/rdpdr.c /tmp/rdesktop-1.8.0-usbrdr/rdpdr.c
--- /tmp/rdesktop-1.8.0/rdpdr.c	2013-06-20 15:56:16.000000000 +0300
+++ /tmp/rdesktop-1.8.0-usbrdr/rdpdr.c	2014-06-16 19:55:04.634098925 +0300
@@ -263,6 +263,8 @@
 	STREAM s;
 	PRINTER *printerinfo;
 
+	usbrdr_lock(USBRDR_LOCK_RDPDR);
+
 	s = channel_init(rdpdr_channel, announcedata_size());
 	out_uint8a(s, magic, 4);
 	out_uint32_le(s, g_num_devices);
@@ -311,6 +313,8 @@
 
 	s_mark_end(s);
 	channel_send(s, rdpdr_channel);
+
+	usbrdr_unlock(USBRDR_LOCK_RDPDR);
 }
 
 void
@@ -323,6 +327,8 @@
 #ifdef WITH_SCARD
 	scard_lock(SCARD_LOCK_RDPDR);
 #endif
+	usbrdr_lock(USBRDR_LOCK_RDPDR);
+
 	s = channel_init(rdpdr_channel, 20 + length);
 	out_uint8a(s, magic, 4);
 	out_uint32_le(s, device);
@@ -337,6 +343,8 @@
 	/* hexdump(s->channel_hdr + 8, s->end - s->channel_hdr - 8); */
 #endif
 	channel_send(s, rdpdr_channel);
+
+	usbrdr_unlock(USBRDR_LOCK_RDPDR);
 #ifdef WITH_SCARD
 	scard_unlock(SCARD_LOCK_RDPDR);
 #endif
diff -u /tmp/rdesktop-1.8.0/secure.c /tmp/rdesktop-1.8.0-usbrdr/secure.c
--- /tmp/rdesktop-1.8.0/secure.c	2013-06-20 15:56:16.000000000 +0300
+++ /tmp/rdesktop-1.8.0-usbrdr/secure.c	2014-06-16 19:55:04.638098925 +0300
@@ -336,6 +336,7 @@
 #ifdef WITH_SCARD
 	scard_lock(SCARD_LOCK_SEC);
 #endif
+	usbrdr_lock(USBRDR_LOCK_SEC);
 
 	s_pop_layer(s, sec_hdr);
 	if ((!g_licence_issued && !g_licence_error_result) || (flags & SEC_ENCRYPT))
@@ -357,6 +358,7 @@
 
 	mcs_send_to_channel(s, channel);
 
+	usbrdr_unlock(USBRDR_LOCK_SEC);
 #ifdef WITH_SCARD
 	scard_unlock(SCARD_LOCK_SEC);
 #endif
@@ -367,7 +369,11 @@
 void
 sec_send(STREAM s, uint32 flags)
 {
+	usbrdr_lock(USBRDR_LOCK_CHANNEL);
+
 	sec_send_to_channel(s, flags, MCS_GLOBAL_CHANNEL);
+
+	usbrdr_unlock(USBRDR_LOCK_CHANNEL);
 }
 
 
diff -u /tmp/rdesktop-1.8.0/tcp.c /tmp/rdesktop-1.8.0-usbrdr/tcp.c
--- /tmp/rdesktop-1.8.0/tcp.c	2013-06-28 14:20:08.000000000 +0300
+++ /tmp/rdesktop-1.8.0-usbrdr/tcp.c	2014-06-16 20:04:32.338110482 +0300
@@ -51,11 +51,13 @@
 #define INADDR_NONE ((unsigned long) -1)
 #endif
 
-#ifdef WITH_SCARD
+/*#ifdef WITH_SCARD
 #define STREAM_COUNT 8
 #else
 #define STREAM_COUNT 1
 #endif
+*/
+#define STREAM_COUNT 16
 
 static RD_BOOL g_ssl_initialized = False;
 static SSL *g_ssl = NULL;
@@ -93,14 +95,28 @@
 STREAM
 tcp_init(uint32 maxlen)
 {
-	static int cur_stream_id = 0;
+	int cur_stream_id;
 	STREAM result = NULL;
 
 #ifdef WITH_SCARD
 	scard_lock(SCARD_LOCK_TCP);
 #endif
-	result = &g_out[cur_stream_id];
-	cur_stream_id = (cur_stream_id + 1) % STREAM_COUNT;
+	usbrdr_lock(USBRDR_LOCK_TCP);
+
+	for(cur_stream_id = 1; cur_stream_id < STREAM_COUNT; cur_stream_id++)
+	{
+		result = &g_out[cur_stream_id];
+		if (result->occupied)
+		{       result = NULL;
+			continue;
+		}
+	}
+
+	if (result == NULL)
+	{
+		DEBUG(("!!! Can't find free stream. Use zero-based STREAM struct\n"));
+		result = &g_out[0];
+	}
 
 	if (maxlen > result->size)
 	{
@@ -108,8 +124,11 @@
 		result->size = maxlen;
 	}
 
+	result->occupied = 1;
 	result->p = result->data;
 	result->end = result->data + result->size;
+
+	usbrdr_unlock(USBRDR_LOCK_TCP);
 #ifdef WITH_SCARD
 	scard_unlock(SCARD_LOCK_TCP);
 #endif
@@ -130,6 +149,8 @@
 #ifdef WITH_SCARD
 	scard_lock(SCARD_LOCK_TCP);
 #endif
+	usbrdr_lock(USBRDR_LOCK_TCP);
+
 	while (total < length)
 	{
 		if (g_ssl)
@@ -146,6 +167,7 @@
 				}
 				else
 				{
+					usbrdr_unlock(USBRDR_LOCK_TCP);
 #ifdef WITH_SCARD
 					scard_unlock(SCARD_LOCK_TCP);
 #endif
@@ -168,6 +190,7 @@
 				}
 				else
 				{
+					usbrdr_unlock(USBRDR_LOCK_TCP);
 #ifdef WITH_SCARD
 					scard_unlock(SCARD_LOCK_TCP);
 #endif
@@ -180,6 +203,10 @@
 		}
 		total += sent;
 	}
+
+	s->occupied = 0;
+
+	usbrdr_unlock(USBRDR_LOCK_TCP);
 #ifdef WITH_SCARD
 	scard_unlock(SCARD_LOCK_TCP);
 #endif
@@ -602,6 +629,7 @@
 		g_out[i].sec_hdr = NULL;
 		g_out[i].rdp_hdr = NULL;
 		g_out[i].channel_hdr = NULL;
+		g_out[i].occupied = 0;
 	}
 }
 
