--- ../../alsa-kernel/pci/riptide/riptide.c	2009-05-26 14:05:54.000000000 +0200
+++ riptide.c	2009-05-26 14:17:52.000000000 +0200
@@ -1,3 +1,4 @@
+#include "adriver.h"
 /*
  *   Driver for the Conexant Riptide Soundchip
  *
@@ -1237,7 +1238,12 @@ static int try_to_load_firmware(struct c
 	snd_printdd("Writing Firmware\n");
 	if (!chip->fw_entry) {
 		err = request_firmware(&chip->fw_entry, "riptide.hex",
-				       &chip->pci->dev);
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 5, 0)
+				       pci_name(chip->pci)
+#else
+				       &chip->pci->dev
+#endif
+				       );
 		if (err) {
 			snd_printk(KERN_ERR
 				   "Riptide: Firmware not available %d\n", err);
@@ -2212,3 +2218,5 @@ static void __exit alsa_card_riptide_exi
 
 module_init(alsa_card_riptide_init);
 module_exit(alsa_card_riptide_exit);
+
+EXPORT_NO_SYMBOLS;
