diff -u -N -r /tmp/FreeRDP-FreeRDP-cd43ba1/channels/CMakeLists.txt /tmp/FreeRDP-FreeRDP-cd43ba1-usbrdr/channels/CMakeLists.txt
--- /tmp/FreeRDP-FreeRDP-cd43ba1/channels/CMakeLists.txt	2012-01-16 19:52:14.000000000 +0200
+++ /tmp/FreeRDP-FreeRDP-cd43ba1-usbrdr/channels/CMakeLists.txt	2012-02-22 14:38:14.000000000 +0200
@@ -23,4 +23,5 @@
 add_subdirectory(rdpdr)
 add_subdirectory(rail)
 add_subdirectory(rdpsnd)
+add_subdirectory(usbrdr)
 
