--- ../../alsa-kernel/pci/echoaudio/indigoiox.c	2009-03-01 15:26:34.000000000 +0100
+++ indigoiox.c	2009-03-01 15:32:56.000000000 +0100
@@ -1,3 +1,9 @@
+#include "adriver.h"
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 5, 0)
+#define pci_device(chip) pci_name(chip->pci)
+#endif
+
 /*
  *  ALSA driver for Echoaudio soundcards.
  *  Copyright (C) 2009 Giuliano Pochini <pochini@shiny.it>
@@ -107,3 +113,5 @@ static struct snd_pcm_hardware pcm_hardw
 #include "echoaudio_dsp.c"
 #include "echoaudio.c"
 
+
+EXPORT_NO_SYMBOLS;
