--- ../../alsa-kernel/include/ac97_codec.h	2009-12-10 14:43:03.000000000 +0100
+++ ac97_codec.h	2009-12-16 18:53:25.000000000 +0100
@@ -645,10 +645,12 @@
 /* ad hoc AC97 device driver access */
 extern struct bus_type ac97_bus_type;
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 5, 0)
 /* AC97 platform_data adding function */
 static inline void snd_ac97_dev_add_pdata(struct snd_ac97 *ac97, void *data)
 {
 	ac97->dev.platform_data = data;
 }
+#endif
 
 #endif /* __SOUND_AC97_CODEC_H */
