--- ../../alsa-kernel/include/pcm_oss.h	2010-02-05 08:53:31.000000000 +0100
+++ pcm_oss.h	2011-03-04 12:24:57.000000000 +0100
@@ -61,7 +61,12 @@
 	struct snd_pcm_plugin *plugin_first;
 	struct snd_pcm_plugin *plugin_last;
 #endif
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 33) || \
+	!defined(__GENKSYMS__)
 	unsigned int prev_hw_ptr_period;
+#else
+	unsigned int prev_hw_ptr_interrupt;
+#endif
 };
 
 struct snd_pcm_oss_file {
